gpii/node_modules/lifecycleManager/src/LifecycleManager.js

```diff
@@ -490,33 +517,82 @@ var gpii = fluid.registerNamespace("gpii");
     };
 
     /**
+     * The lifecycleManager queue is used to hold the high-level actions that needs to happen,
+     * such as starting the login process, starting logout process, starting the update process.
+     * Each item in the queue should have the format: { func: <functionToCall>, args: <Arguments> }
+     * Where <functionToCall> is a function that returns a promise for when it's complete (including side-effects)
+     * and the arguments are the arguments that should be passed to the function.
+     * The queue is run sequentially, and an item is considered "done" once the promise returned by its
+     * function is resolved.
+     */
+    gpii.lifecycleManager.addToQueue = function (that, queue, item) {
+        item.promise = fluid.promise();
+        queue.push(item);
+        if (!that.isProcessingQueue) {
```

Comment:
This logic is identical to line 555-559 of processQueue and should be factored into its own function